### PR TITLE
Pin packages

### PR DESCRIPTION
**Pull request by adriendelsalle**
Description
--

Add capability to pin packages:
- from rc files
- from `<prefix>/conda-meta/pinned` file

Also auto-pin python version if already installed in prefix and specs does not include python at least pinned with minor version.

Add `--no-pin` flag to `create` and `install` CLI sub-commands to disable pinned specs from files (rc or prefix located).

### CODE REVIEW

File: src/pinning.cpp [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+namespace mamba
+{
+    void pin_python_spec(const PrefixData& prefix_data, std::vector<std::string>& specs)
+    {
+        bool py_found = false;
+        std::string py_spec = "";
+
+        for (auto spec : specs)
+        {
+            if (spec.find("python") != std::string::npos)
```
Comment (wolfv) [inline src/pinning.cpp:21]: I am not sure this works. Can we not use the MatchSpec facilities for proper parsing?

**Comment by wolfv** [comment]
Note:

this is how it's done from the Python side: https://github.com/mamba-org/mamba/blob/53e31e4c4973e5b715ae00f6416907be2cae9fd3/mamba/mamba.py#L537-L538

File: src/pinning.cpp (hunk 1 above)
Comment (adriendelsalle) [reply src/pinning.cpp]: fixed!

**Comment by adriendelsalle** [comment]
> Note:
> 
> this is how it's done from the Python side:
> 
> https://github.com/mamba-org/mamba/blob/53e31e4c4973e5b715ae00f6416907be2cae9fd3/mamba/mamba.py#L537-L538

Thanks @wolfv ! I missed that point and now add pins to the solver.

File: src/micromamba/main.cpp [hunk 2]
```diff
@@ -599,6 +602,18 @@ install_specs(const std::vector<std::string>& specs, bool create_env = false, in
     MSolver solver(pool, { { SOLVER_FLAG_ALLOW_DOWNGRADE, 1 } });
     solver.add_jobs(create_options.specs, SOLVER_INSTALL);
 
+    if (!create_options.no_pin)
+    {
+        solver.add_pins(file_pins(prefix_data.path() / "conda-meta/pinned"));
```
Comment (wolfv) [inline src/micromamba/main.cpp:607]: this is not cross-platform. 

Use `...path() / "conda-meta" / "pinned"`

File: src/pinning.cpp [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+namespace mamba
+{
+    std::string python_pin(const PrefixData& prefix_data, const std::vector<std::string>& specs)
+    {
+        std::string pin = "";
+        std::string py_version;
+        MatchSpec ms;
+
+        try
```
Comment (wolfv) [inline src/pinning.cpp:20]: hmm, I am more a fan of (from stackoverflow):

```
MyMapType::const_iterator iter = mymap.find(key);

if(iter != mymap.end())
{
    // iter is item pair in the map. The value will be accessible as `iter->second`.
}
```

**Comment by wolfv** [comment]
this code looks great. Two nitpicks, then it's ready to be merged!

File: src/micromamba/main.cpp (hunk 2 above)
Comment (adriendelsalle) [reply src/micromamba/main.cpp:607]: nice catch, thanks!

File: src/pinning.cpp (hunk 3 above)
Comment (adriendelsalle) [reply src/pinning.cpp]: Thanks again for the suggestion!

**Comment by wolfv** [comment]
cool. Maybe we should make the minor-pin behavior configurable in the future. 

It could be interesting / useful for other programming languages well (e.g. Ruby 2.X / 3.X, Lua, ...).

**Pull request merged**